How-to-upgrade-review-version.md

```diff
@@ -0,0 +1,399 @@
+
+## Fix references to types in dependencies
+
+In analyze-scope.js, I see quite a few errors about missing estree
+types like Identifier and ClassDeclaration. That's weird, because
+those types *do* exist in estree. The problem is that they're not
+imported. I'd like to write
+
+```js
+import { Identifier, ClassDeclaration } from "estree";
+```
+
+But this doesn't work because those are types, not values. The import
+will fail at runtime. Instead, you need to use an *import type*. An
+import type is just like a dynamic import, except that it's used as a
+type. So, just like you could write:
+
+```js
+const fs = import("fs");
+```
+
+to dynamically import `fs`, you can write:
+
+```ts
+var id: import("estree").Identifier = ...
+```
+
+to import the type `Identifier` without an import statement. And,
+because it's inconvenient to repeat `import` all over the place, you
+usually want to write a `typedef`:
+
+```js
+/** @typedef {import("estree").Identifier} Identifier */
+```
+
+## Add missing types in your own code
+
+Fixing these types still leaves a lot of undefined types in
+analyze-scope.js. The types *look* like estree types, but they're
+prefixed with TS-, like TSTypeAnnotation and TSTypeQuery. It turns out
+that these types are specific to typescript-eslint-query. So we'll
+have to define these types ourselves. To start, I defined a lot
+more typedefs with any. This got rid of the errors:
+
+```js
+/** @typedef {any} TSTypeQuery */
+// etc ...
+```
+
+Then I changed the typedefs one by one to 'unknown', and went looking
+for errors that popped up:
+
+```js
+/** @typedef {unknown} TSTypeQuery */
+```
+
+Turns out that only one usage exists, in the method TSQualifiedName:
+
+```js
+    /**
+    * Create reference objects for the object part. (This is `obj.prop`)
+    * @param {TSTypeQuery} node The TSTypeQuery node to visit.
+    * @returns {void}
+    */
+    TSQualifiedName(node) {
+        this.visit(node.left);
+    }
+```
+
+TSTypeQuery is supposed to have a left property, so I changed from
+`unknown` to `{ left: unknown }`:
+
+```js
+/** @typedef {{ left: unknown }} TSTypeQuery */
+```
+
+With little knowledge of typescript-eslint-parser, I can only use the
+code itself to figure out the types, but of course if you work on a
+project, you'll have a better idea of what the types should be.
+
+(and use them)
+(and decide where to store them)
+(CHECK YOUR WORK)
+
+## Work around missing types
+
+Sometimes you'll find that one of your dependencies has no `@types`
+package at all. You are free to define types and contribute them to
+Definitely Typed, of course, but you usually need a quick way to work
+around missing dependencies. The easiest way is to add your own
+typings file to hold workarounds. For example, the next error left in
+analyze-scope.js is on:
+
+```js
+new PatternVisitor(this.options, node, callback)
+```
+
+It says that PatternVisitor expects 0 arguments but got 3. Why did it
```

Comment:
Where is `PatternVisitor`? Is this the source code or an external library? A source link may help so I don't get lost